web/client/epics/map.js

```diff
@@ -214,11 +222,48 @@ const checkMapPermissions = (action$, {getState = () => {} }) =>
             return loadMapInfo(ConfigUtils.getConfigProp("geoStoreUrl") + "extjs/resource/" + mapId, mapId);
         });
 
+const compareMapChanges = (action$, { getState = () => {} }) =>
+    action$
+        .ofType(CHECK_MAP_CHANGES)
+        .switchMap(({ action, source }) => {
+
+            const mapId = getState().map && getState().map.present.mapId;
+            const { view } = getState().feedbackMask;
+
+            if (!mapId && view !== 'viewer') {
+                action && action();
+                return Rx.Observable.empty();
+            }
+
+            const url = ConfigUtils.getConfigUrl({ mapId }).configUrl;
+            return Rx.Observable.defer(() => axios.get(url))
+                .switchMap(({ data: originalMap }) => {
+                    const updatedMap = MapUtils.saveMapConfiguration(
+                        mapSelector(getState()),
+                        layersSelector(getState()),
+                        groupsSelector(getState()),
+                        backgroundListSelector(getState()),
+                        textSearchConfigSelector(getState()),
+                        mapOptionsToSaveSelector(getState())
+                    );
+                    const isEqual = MapUtils.compareMapChanges(originalMap, updatedMap);
+                    if (!isEqual) {
+                        return Rx.Observable.of(
+                            setControlProperty('unsavedMap', 'enabled', true, false),
+                            setControlProperty('unsavedMap', 'source', source, false)
+                        );
+                    }
+                    action && action();
```

Comment:
Could this be better?

return action ? Rx.Observable.of(action) : Rx.Observable.empty();

Reply:
agree i fixed it